File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Go/da16033ce77af907c76da5f666b92aea.go
package main

import (
	"fmt"
	"math"
	"sort"
)

type ll int64

var pi = 2 * math.Acos(0.0)

var id1, testCase = 0, 1

var mainArray [][]ll
var dpTable [][]ll

func functionName() {
	var arrayNumber, operationNumber int
	fmt.Scanf("%d %d", &arrayNumber, &operationNumber)

	id3 := make([][]ll, operationNumber)
	for i := range id3 {
		id3[i] = make([]ll, arrayNumber)
	}

	mainArray = make([][]ll, arrayNumber)
	for i := range mainArray {
		mainArray[i] = make([]ll, operationNumber)
	}

	maximum := ll(0)
	dpTable = make([][]ll, arrayNumber)
	for i := range dpTable {
		dpTable[i] = make([]ll, operationNumber+1)
		for j := range dpTable[i] {
			dpTable[i][j] = maximum
		}
	}

	for i := 0; i < arrayNumber; i++ {
		var elementNumber int
		fmt.Scanf("%d", &elementNumber)
		for j := 0; j < elementNumber; j++ {
			var element ll
			fmt.Scanf("%d", &element)
			if j < operationNumber {
				if j > 0 {
					mainArray[i][j] = element + mainArray[i][j-1]
				} else {
					mainArray[i][j] = element
				}
				id3[j][i] = mainArray[i][j]
			}
		}
	}

	for i := 0; i < operationNumber; i++ {
		sort.Slice(id3[i], func(a, b int) bool {
			return id3[i][a] > id3[i][b]
		})
	}

	for i := 0; i < arrayNumber; i++ {
		for j := 0; j < operationNumber; j++ {
			if mainArray[i][j] < id3[j][min(operationNumber/(j+1), arrayNumber-1)] {
				mainArray[i][j] = 0
			}
		}
	}

	for id0 := 0; id0 < arrayNumber; id0++ {
		if id0 > 0 {
			dpTable[id0] = dpTable[id0-1]
		}

		for id2 := 0; id2 < operationNumber; id2++ {
			if mainArray[id0][id2] != 0 {
				for operationLeft, old := operationNumber, operationNumber-(id2+1); operationLeft > id2; operationLeft, old = operationLeft-1, old-1 {
					if id0 == 0 {
						dpTable[id0][operationLeft] = max(dpTable[id0][operationLeft], mainArray[id0][id2])
					} else {
						dpTable[id0][operationLeft] = max(dpTable[id0][operationLeft], mainArray[id0][id2]+dpTable[id0-1][old])
					}
				}
			}
		}
	}

	answer := dpTable[arrayNumber-1][operationNumber]
	fmt.Println(answer)
}

func functionNameWithParams(v1, v2, v3 int) {}

func functionNameWithString(line string) {}

type initialWorks struct{}

func (iw *initialWorks) withTestCaseNumber() {
	fmt.Scanf("%d", &id1)

	for testCase <= id1 {
		functionName()
		testCase++
	}
}

func (iw *initialWorks) singleIteration() {
	functionName()
}

func (iw *initialWorks) withoutTestCaseNumber() {
	var v1, v2, v3 int
	for fmt.Scanf("%d %d %d", &v1, &v2, &v3) == 3 && v1 != 0 {
		functionNameWithParams(v1, v2, v3)
		testCase++
	}
}

func (iw *initialWorks) eoof() {
	var id4 string
	for {
		if _, err := fmt.Scanf("%[^\n]", &id4); err != nil {
			break
		}
		fmt.Scanf("%*c") // consume newline

		functionNameWithString(id4)
		testCase++
	}
}

func main() {
	var a initialWorks
	a.singleIteration()
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b ll) ll {
	if a > b {
		return a
	}
	return b
}

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Python/da16033ce77af907c76da5f666b92aea.py
import sys
import math
from collections import defaultdict

pi = 2 * math.acos(0.0)

id1 = 0
test_case = 1

main_array = []
dp_table = []

def function_name():
    global main_array, dp_table
    array_number, operation_number = map(int, sys.stdin.readline().split())
    
    id3 = [[0] * array_number for _ in range(operation_number)]
    main_array = [[0] * operation_number for _ in range(array_number)]
    
    maximum = 0
    dp_table = [[maximum] * (operation_number + 1) for _ in range(array_number)]

    for i in range(array_number):
        element_number = int(sys.stdin.readline().strip())
        for j in range(element_number):
            element = int(sys.stdin.readline().strip())
            if j < operation_number:
                main_array[i][j] = element + (j > 0) * main_array[i][j - 1]
                id3[j][i] = main_array[i][j]

    for i in range(operation_number):
        id3[i].sort(reverse=True)

    for i in range(array_number):
        for j in range(operation_number):
            if main_array[i][j] < id3[j][min(operation_number // (j + 1), array_number - 1)]:
                main_array[i][j] = 0

    for id0 in range(array_number):
        if id0 > 0:
            dp_table[id0] = dp_table[id0 - 1][:]

        for id2 in range(operation_number):
            if main_array[id0][id2] != 0:
                for operation_left in range(operation_number, id2, -1):
                    old = operation_number - (id2 + 1)
                    if id0 == 0:
                        dp_table[id0][operation_left] = max(dp_table[id0][operation_left], main_array[id0][id2])
                    else:
                        dp_table[id0][operation_left] = max(dp_table[id0][operation_left], main_array[id0][id2] + dp_table[id0 - 1][old])

    answer = dp_table[array_number - 1][operation_number]
    print(answer)

def function_name_v1(v1, v2, v3):
    pass

def function_name_line(line):
    pass

class InitialWorks:
    def with_test_case_number(self):
        global id1, test_case
        id1 = int(sys.stdin.readline().strip())

        while test_case <= id1:
            function_name()
            test_case += 1

    def single_iteration(self):
        function_name()

    def without_test_case_number(self):
        global test_case
        while True:
            line = sys.stdin.readline().strip()
            if not line:
                break
            v1, v2, v3 = map(int, line.split())
            if v1 == 0:
                break
            function_name_v1(v1, v2, v3)
            test_case += 1

    def eoof(self):
        global test_case
        while True:
            try:
                line = input()
                function_name_line(line)
                test_case += 1
            except EOFError:
                break

if __name__ == "__main__":
    initial_works = InitialWorks()
    initial_works.single_iteration()
